File: backend/src/main/java/com/chatbot/modules/address/model/Address.java
package com.chatbot.modules.address.model;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.Filter;
import org.hibernate.annotations.FilterDef;
import org.hibernate.annotations.ParamDef;

@Entity
@Table(
    name = "addresses",
    indexes = {
        @Index(name = "idx_address_tenant", columnList = "tenant_id"),
        @Index(name = "idx_address_owner", columnList = "tenant_id, owner_type, owner_id")
    }
)
// --- THÊM PHẦN NÀY ---
@FilterDef(
    name = "tenantFilter",
    parameters = @ParamDef(name = "tenantId", type = Long.class)
)
@Filter(name = "tenantFilter", condition = "tenant_id = :tenantId")
// ---------------------
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Address {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /**
     * Tenant sở hữu address
     */
    @Column(name = "tenant_id", nullable = false)
    private Long tenantId;

    /**
     * Loại owner: USER, END_USER, ORDER, SHOP...
     */
    @Enumerated(EnumType.STRING)
    @Column(name = "owner_type", nullable = false, length = 30)
    private OwnerType ownerType;

    /**
     * ID của owner tương ứng với ownerType
     */
    @Column(name = "owner_id", nullable = false)
    private Long ownerId;

    // ===== Địa chỉ =====
    @Column(name = "house_number", length = 50)
    private String houseNumber;

    @Column(length = 255)
    private String street;

    @Column(length = 255)
    private String ward;

    @Column(length = 255)
    private String district;

    @Column(length = 255)
    private String province;

    @Column(length = 255)
    private String country;

    @Builder.Default
    @Column(name = "is_default", nullable = false)
    private boolean isDefault = false;
    
    // Thêm setter cho compatibility
    public void setDefault(boolean isDefault) {
        this.isDefault = isDefault;
    }
    
    public boolean isDefault() {
        return this.isDefault;
    }
}

File: frontend-new/src/types/address.ts
// types/address.ts

/**
 * Khớp hoàn toàn với OwnerType.java backend
 * Chỉ có USER và TENANT, không có END_USER/ORDER/SHOP
 */
export type OwnerType = 'USER' | 'TENANT';

/**
 * Khớp với AddressEmbedded.java - Dùng cho Tenant/User
 * Backend chỉ dùng model này, không dùng Address table
 */
export interface AddressEmbedded {
  houseNumber?: string;
  street?: string;
  ward?: string;
  district?: string;
  province?: string;
  country?: string;
  fullAddress?: string;
}

// ========================
// HELPER TYPES
// ========================

/**
 * Type cho address của Tenant/User (dùng AddressEmbedded)
 */
export type TenantAddress = AddressEmbedded;
export type UserAddress = AddressEmbedded;

// ========================
// LEGACY TYPES (Để giữ compatibility nhưng sẽ deprecated)
// ========================

/**
 * @deprecated Dùng AddressEmbedded thay thế
 * Backend không sử dụng Address table cho User/Tenant
 */
export interface Address {
  id: number;
  tenantId: number;
  ownerType: OwnerType;
  ownerId: number;
  houseNumber?: string;
  street?: string;
  ward?: string;
  district?: string;
  province?: string;
  country?: string;
  isDefault: boolean; // Cần thiết cho nhiều addresses
}

/**
 * @deprecated Dùng AddressEmbedded thay thế
 */
export interface AddressResponseDTO {
  id: number;
  fullAddress: string;
  isDefault: boolean;
}

/**
 * @deprecated Dùng AddressEmbedded thay thế
 */
export interface AddressDetailResponseDTO extends AddressResponseDTO {
  tenantId: number;
  ownerType: OwnerType;
  ownerId: number;

  houseNumber: string | null;
  street: string;
  ward: string;
  district: string | null;
  province: string | null;
  country: string | null;
}

/**
 * @deprecated Dùng AddressEmbedded thay thế
 */
export interface AddressRequestDTO {
  ownerType: OwnerType;
  ownerId: number;
  
  street: string;
  houseNumber?: string;
  ward: string;
  district?: string;
  province?: string;
  country?: string;
  isDefault: boolean; // Cần thiết cho nhiều addresses
}

// ========================
// NOTE: Backend chỉ hỗ trợ USER và TENANT addresses
// Không có END_USER, ORDER, SHOP như comment cũ
// ========================
